### PR TITLE
[RaySGD] Add tqdm logging to TorchTrainer

### PR DESCRIPTION
**Pull request by maximsmol**
## Why are these changes needed?

Currently the only way to log epoch progress is to print to stdout, which ray will send to monitor. This causes a separate line for each report and looks awful. This PR implements a pretty tqdm progress bar for `train`.

## Related issue number
N/A

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by maximsmol** [comment]
CC: @richardliaw

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -1,22 +1,43 @@
 ---
 name: Bug report
-about: Report a bug on the latest version of Ray.
```
Comment (richardliaw) [inline .github/ISSUE_TEMPLATE/bug_report.md]: revert

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 2]
```diff
@@ -1,12 +1,20 @@
 ---
```
Comment (richardliaw) [inline .github/ISSUE_TEMPLATE/feature_request.md]: revert?

File: python/ray/util/sgd/torch/batch_logs_reporter.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import ray
+
+
+@ray.remote(num_cpus=0)
+class BatchLogsReporter:
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/batch_logs_reporter.py]: Should this be private? `_Batch..`

File: python/ray/util/sgd/torch/training_operator.py [hunk 4]
```diff
@@ -318,6 +368,25 @@ def optimizers(self):
...
+        """
+        Data loader for the validation dataset created by the data_creator.
+        """
+        return self._train_loader
+
+    @property
+    def validation_loader(self):
+        """
+        Data loader for the train dataset created by the data_creator.
+        """
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: These docstrings should be one-line

File: python/ray/util/sgd/torch/training_operator.py [hunk 5]
```diff
@@ -154,6 +197,13 @@ def train_epoch(self, ...):
 
         return metric_meters.summary()
 
+    def forward(self, features, target):
+        output = self.model(features)
+        return self.criterion(output, target)
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: Can we remove this? this isn't fundamental right?

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 1 above)
Comment (maximsmol) [reply .github/ISSUE_TEMPLATE/bug_report.md]: I'll hard-reset everything outside of `util/sgd` git seems to have messed a bunch of stuff up

File: python/ray/util/sgd/torch/batch_logs_reporter.py (hunk 3 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/batch_logs_reporter.py]: It's not exported from the main file, but sure --- will do

File: python/ray/util/sgd/torch/training_operator.py (hunk 5 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/training_operator.py]: Probably should remove from this PR, but overall I think it's a good idea to add it at some point. I'll motivate in another PR

File: python/ray/util/sgd/torch/training_operator.py [hunk 6]
```diff
@@ -318,6 +367,25 @@ def optimizers(self):
         """List of optimizers created by the ``optimizer_creator``."""
         return self._optimizers
 
+    @property
+    def train_loader(self):
+        """
+        Data loader for the validation dataset created by the ``data_creator``.
+        """
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: BTW, if there is only one sentence, it should be on one line - 
```suggestion
        """Data loader for the validation dataset created by the ``data_creator``."""
```

https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: python/ray/util/sgd/torch/training_operator.py [hunk 7]
```diff
@@ -318,6 +367,25 @@ def optimizers(self):
...
+        return self._train_loader
+
+    @property
+    def validation_loader(self):
+        """
+        Data loader for the train dataset created by the ``data_creator``.
+        """
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: same as above

File: python/ray/util/sgd/torch/training_operator.py [hunk 8]
```diff
@@ -142,6 +177,14 @@ def train_epoch(self, ...):
             batch_info.update(info)
             metrics = self.train_batch(batch, batch_info=batch_info)
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: Ideally, here we just have `self.process_batch(metrics)`. This makes it easier for someone to extend the operator while not losing the functionality of the package.

File: python/ray/util/sgd/torch/training_operator.py [hunk 9]
```diff
@@ -79,6 +92,20 @@ def __init__(self,
...
+        return ray.get(self._batch_logs_reporter._send_setup.remote(data))
+
+    def _send_batch_logs(self, data):
+        cur_time = time.monotonic()
+        if cur_time - self._last_batch_logs_send < BATCH_LOGS_RATE_LIMIT:
+            return
+
+        self._last_batch_logs_send = cur_time
+        return ray.get(self._batch_logs_reporter._send.remote(data))
+
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py]: As mentioned offline, I was thinking that all this logic be hidden in a separate handler or class. 

Maybe one good rule of thumb is for the operator not to need to call Ray primitives (not sure)

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 10]
```diff
@@ -359,8 +365,37 @@ def train(self,
...
+                        desc = "{}e".format(info["epoch_idx"] + 1)
+
+                batch_pbar = tqdm(
+                    total=n, desc=desc, unit="batch", leave=False)
+                return
+
+            if packet["packet_type"] == "batch_logs":
+                batch_pbar.n = packet["batch_idx"] + 1
+                batch_pbar.set_postfix(packet["pbar_metrics"])
+
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py]: Let's isolate this to another class specific to tqdm?

**Comment by richardliaw** [comment]
This is a first pass at what I'm thinking:

```
class Trainer
    def train(self, tqdm=True):
        if tqdm:
            handler = TqdmHandler()
            self.handlers.append(handler)
            self.workers[0].register_handler(handler.create_reporter())


    def _train_epoch(self, ...):
        try:
            unfinished = run_all_training() 

            while len(unfinished) > 0:
                finished, unfinished = ray.wait(
                    unfinished, timeout=BATCH_LOGS_RATE_LIMIT)

                # throw errors on agent failure
                finished = ray.get(finished)

                for handler in self.handlers:
                    hander.step()

            return True, worker_trains
        except RayActorError as exc:
            logger.exception(str(exc))
        return False, worker_trains



class TqdmHandler():
    def __init__():
        self.batch_reporter = BatchLogReporter.remote()

    def step(self):
        metrics = ray.get(self.batch_reporter._read_setup.remote()))
        self.batch_logs_handler(metrics)


    def _batch_logs_handler(self, metrics):
        pass

    def create_reporter(self):
        return TqdmReporter(self.batch_reporter)

@ray.remote
class BatchLogReporter():
    ...

class Operator()
    def register_handler(self, handler):
        self.handlers += [handler]

    def train():
        for batch in dataloader:
            metrics = train_batch(batch)
            self.process_batch_metrics(metrics)

    def process_batch_metrics(self, metrics)
        for c in self.handlers:
            c.process(metrics) 
     

TqdmReporter():
    def __init__(self):

    def process(self, metrics):
        if self.should_send():
            self.actor.send.remote()
```

File: python/ray/util/sgd/torch/training_operator.py (hunk 6 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/training_operator.py]: It does not fit in 79 characters

File: python/ray/util/sgd/torch/training_operator.py (hunk 7 above)
Comment (maximsmol) [reply python/ray/util/sgd/torch/training_operator.py]: also will not fit :PP

**Review by richardliaw — APPROVED** [review]

**Pull request merged**